pandasai/ee/vectorstores/lanceDB.py

```diff
@@ -0,0 +1,411 @@
+import uuid
+from typing import Callable, Iterable, List, Optional, Union
+
+import lancedb
+import pandas as pd
+from lancedb.embeddings import get_registry
+from lancedb.pydantic import LanceModel, Vector
+from sentence_transformers import SentenceTransformer
+
+from pandasai.helpers.logger import Logger
+from pandasai.vectorstores.vectorstore import VectorStore
+
+
+class LanceDB(VectorStore):
+    """
+    Implementation of LanceDB vector store
+    """
+
+    _logger: Logger
+
+    def __init__(
+        self,
+        table_name: str = "pandasai",
+        embedding_function: Optional[Callable[[List[str]], List[float]]] = None,
+        persist_path: Optional[str] = "/tmp/lancedb",
+        max_samples: int = 1,
+        similary_threshold: int = 1.5,
+        logger: Optional[Logger] = None,
+    ) -> None:
+        self._logger = logger or Logger()
+        self._max_samples = max_samples
+        self._similarity_threshold = similary_threshold
+        self._persist_directory = persist_path
+
+        # Initialize LanceDB database
+        self._db = lancedb.connect(self._persist_directory)
+        # Embedding function
+        self._embedding_function = embedding_function
+
+        if self._embedding_function is None:
+            embedding_function = (
+                get_registry()
+                .get("sentence-transformers")
+                .create(name="BAAI/bge-small-en-v1.5", device="cpu")
+            )
+
+            # schema for qa-pairs
+            class QA_pairs(LanceModel):
+                id: str
+                qa: str = embedding_function.SourceField()
+                metadata: str
+                vector: Vector(
+                    embedding_function.ndims()
+                ) = embedding_function.VectorField()
+
+            # schema for docs
+            class Docs(LanceModel):
+                id: str
+                doc: str = embedding_function.SourceField()
+                metadata: str
+                vector: Vector(
+                    embedding_function.ndims()
+                ) = embedding_function.VectorField()
+
+        else:
+            # schema for qa-pairs
+            class QA_pairs(LanceModel):
```

Comment:
These classes can be declared on outside LanceDB and on the basis of check we set vector value instead. These pairs  and docs are needed for all cases.

Reply:
Done